library/alloc/src/collections/btree/navigate.rs

```diff
@@ -1,7 +1,14 @@
+use core::intrinsics;
+use core::mem;
 use core::ptr;
 
 use super::node::{marker, ForceResult::*, Handle, NodeRef};
-use super::unwrap_unchecked;
+
+/// `Option::unwrap` without the promise to panic if the option contains no value.
+/// Instead, the entire process would be aborted.
+fn unwrap_or_abort<T>(val: Option<T>) -> T {
+    val.unwrap_or_else(|| intrinsics::abort())
+}
```

Comment:
This is not necessary. Just let unwrap_unchecked panic in debug build, and let PanicGuard guarantee the safety. They do both end up aborting, but with unwrap_unchecked we have stack backtrace.

Reply:
I saw it as an opportunity, not a necessity. The opportunity to avoid an unsafe call and to avoid [the performance backlash of unwrap_unchecked](#74693), although in navigate.rs, benchmarks are not at all clear which of `unwrap`, `unwrap_unchecked` or `unwrap_or_abort` they think is faster.

As to a stack trace, the abort seems better to me, in my experience, on Windows, and I just confirmed on Linux. RUST_BACKTRACE seems completely useless in these double panic situations. The abort gives an illegal instruction, the debugger takes you right to the action at the top of the call stack, easier than if you have that unwinding kind of stuff on top.

Reply:
> The opportunity to avoid an unsafe call 

Calling unwrap_unchecked is not unsafe anymore with PanicGuard.

> RUST_BACKTRACE seems completely useless in these double panic situations. 

Shouldn't be the case. I just confirmed RUST_BACKTRACE will produce backtrace just fine. Backtrace is generally preferred to just aborting when the situation allows, because it gives you more information without debugger.

> avoid the performance backlash of unwrap_unchecked

unwrap_unchecked uses `intrinsics::unreachable_unchecked()`, and it is *supposed* to be faster. The current performance drawback is due to a bug in LLVM (#74615). If that get's fixed, unwrap_unchecked would be much faster than unwrap_or_abort.

Reply:
> Calling unwrap_unchecked is not unsafe anymore with PanicGuard.

I don't believe you. If there is nothing to unwrap, our call to `unwrap_unchecked` reaches UB and anything can happen. Not a problem, because the function whose body we're talking about is also tagged unsafe and documented to the caller when this could happen. But still unsafe in my book.

My point was that the call requires an unsafe block, even if there wasn't any UB lurking, so reading the code has a cognitive overhead. You could say that having to understand yet another variation on unwrap also has that overhead. Also, I struggle with the fact/perception that aborting is safe and UB, which includes recovering and saving a user's work, is unsafe. Now I think that aborting, although classified as safe, is far worse than unsafe so `unwrap_or_abort` is worse than `unwrap_unchecked`.

> I just confirmed RUST_BACKTRACE will produce backtrace just fine.

Well in simlle situations, sure. But do you mean in a double panic situation, in liballoc code as compiled by x.py with optimize=false, debug=true, debug-assertions=true?

Reply:
Sorry, I misread your comment, I mean calling replace with is not unsafe. You're correct that's unsafe.

Abort is obvious safe, but it is definitely something to avoid if not necessary. For BTreeMap's case though, we know that the option is not None, so use unwrap_unchecked provide a good chance for LLVM to optimise things out (if there isn't a I-slow bug).

I do mean a double panic situation for backtrace. A very simple demonstration on playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4692ea8a7750c37bc3c55144e5fbb12c.

Reply:
Good idea. I stuffed a slight variation:
```
struct P;
impl Drop for P {
    fn drop(&mut self) {
        panic!();
    }
}

fn main() {
    let _p = P;
    panic!();
}

#[test]
fn test_double_panic() {
    main()
}
```

in library\alloc\tests\btree\extra.rs, hooked it up in mod.rs, and then when I `rustc extra.rs && extra.exe`, I have the trace. But up in the top directory, there is no trace for `python x.py test --stage 0 library/alloc --test-args --test-threads=1 --test-args --format=pretty --test-args test_double_panic`. One of the things it does let out is that the cargo command includes `"--features" "panic-unwind backtrace compiler-builtins-c"`.

Reply:
Did you add `RUST_BACKTRACE=1` and `--test-args --nocapture`?

Reply:
> --nocapture

:man_facepalming: